lib/internal/fs/dir.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+
+const pathModule = require('path');
+const binding = internalBinding('fs');
+const dirBinding = internalBinding('fs_dir');
+const {
+  codes: {
+    ERR_MISSING_ARGS
+  }
+} = require('internal/errors');
+
+const { FSReqCallback } = binding;
+const internalUtil = require('internal/util');
+const {
+  getDirent,
+  getOptions,
+  getValidatedPath,
+  handleErrorFromBinding,
+  makeCallback
+} = require('internal/fs/utils');
+
+const kDirHandle = Symbol('Dir handle');
+const kDirPath = Symbol('Dir path');
+const kDirOptions = Symbol('Dir options');
+const kDirReadP = Symbol('Dir Awaitable Read');
+const kDirCloseP = Symbol('Dir Awaitable Close');
+
+class Dir {
+  constructor(handle, path, options) {
+    if (handle == null) throw new ERR_MISSING_ARGS('handle');
+    this[kDirHandle] = handle;
+    this[kDirPath] = path;
+
+    this[kDirOptions] = getOptions(options, {
+      encoding: 'utf8'
+    });
+
+    this[kDirReadP] = internalUtil.promisify(this.read.bind(this));
+    this[kDirCloseP] = internalUtil.promisify(this.close.bind(this));
+  }
+
+  get path() {
+    return pathModule.toNamespacedPath(this[kDirPath]);
+  }
+
+  read(options, callback) {
+    callback = typeof options === 'function' ? options : callback;
+    if (!callback) {
+      return this[kDirReadP](options);
+    }
+    callback = makeCallback(callback);
+    options = getOptions(options, this[kDirOptions]);
+
+    const req = new FSReqCallback();
+    req.oncomplete = (err, result) => {
+      if (err || result === null) {
+        return callback(err, result);
+      }
+      getDirent(this[kDirPath], result[0], result[1], callback);
+    };
+
+    this[kDirHandle].read(
+      options.encoding,
+      req
+    );
+  }
+
+  readSync(options) {
+    options = getOptions(options, this[kDirOptions]);
+
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].read(
+      options.encoding,
+      undefined,
+      ctx
+    );
+    handleErrorFromBinding(ctx);
+
+    if (result === null) {
+      return result;
+    }
+
+    return getDirent(this[kDirPath], result[0], result[1]);
+  }
+
+  close(callback) {
+    if (!callback) {
+      return this[kDirCloseP]();
+    }
+    callback = makeCallback(callback);
+    const req = new FSReqCallback();
+    req.oncomplete = callback;
+    this[kDirHandle].close(req);
+  }
+
+  closeSync() {
+    const ctx = { path: this[kDirPath] };
+    const result = this[kDirHandle].close(undefined, ctx);
+    handleErrorFromBinding(ctx);
+    return result;
+  }
+}
+
+Dir.prototype[Symbol.asyncIterator] = async function* dirReadAsyncIterGen() {
+  while (true) {
+    const result = await this[kDirReadP]();
+    if (result === null) {
+      this[kDirCloseP]();
```

Comment:
I think you need to wrap the contents of the function in 

```js
try {
  while (true) {
  ...
  }
} finally {
  await this[kDirCloseP]();  // Not sure about how the await works here/whether it’s needed
}
```

and remove this line so that it handles premature returns (`return`, `break`, `throw`) from a `for await (... of dir)` loop well and still closes the underlying resource.

Reply:
```js
class Test {
  async* [Symbol.asyncIterator]() {
    let i = 0;
    try {
      while (true)
        yield i++;
    } finally {
      console.log(`code in finally ran at ${i}!`);
    }
  }
}

async function test1() {
  for await (const number of new Test())
    if (number === 2) break;
}

async function test2() {
  for await (const number of new Test())
    if (number === 5) throw new Error();
}

async function test3() {
  for await (const number of new Test())
    if (number === 7) return;
}

test1();
test2().catch(() => {});
test3();
```

prints

```
code in finally ran at 3!
code in finally ran at 6!
code in finally ran at 8!
```

for me, so I think there’s some kind of notification?

Reply:
I think `yield` sadly doesn’t work because it would add an additional element to the iterable result.

> but this code makes me kinda nervous.

Don’t get me wrong, this is dark magic to me too.

Reply:
I mean, I see that when using `yield`, behaviour could be unexpected – after all, the consumer already dropped out – but with `await`, I don’t think there’s an issue like that? And it does seem to handle rejections from inside the `finally` well, too.

Reply:
@Fishrock123 Can you expand on these “other problems”? I spoke to @​littledan* today and he was also of the opinion that `await this[kDirClosePromise]();` inside a `try` / `finally` here would be the right approach for an async generator function. So if there’s issues with this syntax, I think it’s worth being explicit about what those issues are, because they might require addressing at the language level.

<sub>(* no @​mention because I don’t want to flood people’s inboxes unless necessary)</sub>

Reply:
Ok I'm not really sure what happened but it... works now? Weird.